module/zfs/spa_errlog.c

```diff
@@ -128,91 +249,599 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
-	    &count) == 0)
-		total += count;
-	mutex_exit(&spa->spa_errlog_lock);
+static int
+process_error_block(spa_t *spa, uint64_t head_ds, zbookmark_err_phys_t *zep,
+    uint64_t *count, void *addr, boolean_t only_count)
+{
+	dsl_pool_t *dp = spa->spa_dsl_pool;
+	dsl_pool_config_enter(dp, FTAG);
+	uint64_t top_affected_fs;
 
-	mutex_enter(&spa->spa_errlist_lock);
-	total += avl_numnodes(&spa->spa_errlist_last);
-	total += avl_numnodes(&spa->spa_errlist_scrub);
-	mutex_exit(&spa->spa_errlist_lock);
+	int error = find_top_affected_fs(spa, head_ds, zep, &top_affected_fs);
+	if (error != 0)
+		goto out;
+
+	error = check_filesystem(spa, top_affected_fs, zep, count, addr,
+	    only_count);
+
+out:
+	dsl_pool_config_exit(dp, FTAG);
+	return (error);
+}
+
+static uint64_t
+get_errlog_size(spa_t *spa, uint64_t spa_err_obj)
+{
+	uint64_t total = 0;
+	if (spa_err_obj == 0)
+		return (total);
+
+	zap_cursor_t zc;
+	zap_attribute_t za;
+	for (zap_cursor_init(&zc, spa->spa_meta_objset, spa_err_obj);
+	    zap_cursor_retrieve(&zc, &za) == 0; zap_cursor_advance(&zc)) {
+
+		zap_cursor_t head_ds_cursor;
+		zap_attribute_t head_ds_attr;
+		zbookmark_err_phys_t head_ds_block;
+
+		uint64_t head_ds;
+		name_to_object(za.za_name, &head_ds);
+
+		for (zap_cursor_init(&head_ds_cursor, spa->spa_meta_objset,
+		    za.za_first_integer); zap_cursor_retrieve(&head_ds_cursor,
+		    &head_ds_attr) == 0; zap_cursor_advance(&head_ds_cursor)) {
+
+			name_to_errphys(head_ds_attr.za_name, &head_ds_block);
+			if (process_error_block(spa, head_ds, &head_ds_block,
+			    &total, NULL, B_TRUE) != 0) {
+				zap_cursor_fini(&head_ds_cursor);
+				zap_cursor_fini(&zc);
+				return (total);
+			}
+		}
+		zap_cursor_fini(&head_ds_cursor);
+	}
+	zap_cursor_fini(&zc);
+	return (total);
+}
+
+static uint64_t
+get_errlist_size(spa_t *spa, avl_tree_t *tree)
+{
+	uint64_t total = 0;
+	if (avl_numnodes(tree) == 0)
+		return (total);
 
+	spa_error_entry_t *se;
+	for (se = avl_first(tree); se != NULL; se = AVL_NEXT(tree, se)) {
+		zbookmark_err_phys_t zep;
+		zep.zb_object = se->se_bookmark.zb_object;
+		zep.zb_level = se->se_bookmark.zb_level;
+		zep.zb_blkid = se->se_bookmark.zb_blkid;
+
+		/*
+		 * If we cannot find out the head dataset and birth txg of
+		 * the present error block, we opt not to error out. In the
+		 * next pool sync this information will be retrieved by
+		 * sync_error_list() and written to the on-disk error log.
+		 */
+		uint64_t head_ds_obj;
+		int error = get_head_and_birth_txg(spa, &zep,
+		    se->se_bookmark.zb_objset, &head_ds_obj);
+		if (error != 0)
+			continue;
+
+		(void) process_error_block(spa, head_ds_obj, &zep, &total,
+		    NULL, B_TRUE);
+	}
 	return (total);
 }
+#endif
 
+/*
+ * Return the number of errors currently in the error log.  This is actually the
+ * sum of both the last log and the current log, since we don't know the union
+ * of these logs until we reach userland.
+ */
+void
+spa_get_errlog_size(spa_t *spa, uint64_t *count)
```

Comment:
```diff
- uint64_t spa_get_errlog_size(spa_t *spa)
+ void spa_get_errlog_size(spa_t *spa, uint64_t *count)
```
In every place `spa_get_errlog_size()` is called, `count` is always zeroed first, so there doesn't seem to be a reason for this change.

Reply:
Correct, I will simplify this to directly return the `count`.